[ECMA-262 layering] Refactor import-related host hooks

nicolo-ribaudo opened the pull request:
Hi!

I plan to propose a refactor to the import-related host hooks of ECMA-262, and this PR shows how it would affect HTML.

Currently, ECMA-262 has two host hooks:
- `HostResolveImportedModule(referrer, specifier)`, which synchronously returns the Module Record corresponding to the `(referrer, specifier)` pair;
- `HostImportModuleDynamically(referrer, specifier)`, used for dynamic `import()`, which asynchronously loads the Module Record corresponding to `(referrer, specifier)` and all its dependencies, calls `.Link()`/`.Evaluate()` on it and then "returns" to ECMA-262 by calling `FinishDynamicImport`.

We are working on three ECMA-262 proposals related to modules, and each one of them has needs that are not satisfied by the existing layering:
- [Module Blocks](https://github.com/tc39/proposal-js-module-blocks) allows declaring a module inline. It needs a way to load the dependencies of an already available module record.
- [Import Reflection](https://github.com/tc39/proposal-import-reflection), at least when it comes to JS modules, allows loading a module without loading/linking/evaluating its dependencies. It needs a way to load a module without doing a "full import", and it needs a way to _later_ load/link/evaluate its dependencies.
- ["Layer 0" of Compartments](https://github.com/tc39/proposal-compartments/blob/master/0-module-and-module-source.md) allows implementing custom module loaders, and thus needs to specify how the graph loading algorithm works. This is currently only handled by hosts, for example HTML defines it in [fetch a single module script](https://html.spec.whatwg.org/#fetch-a-single-module-script) and [fetch the descendants of a module script](https://html.spec.whatwg.org/#fetch-the-descendants-of-a-module-script).

Instead of adding multiple new host hooks and duplicating the graph fetching algorithm, we can:
- expose a single host hook, `HostLoadImportedModule(referrer, specifier)`, that loads a single Module Record and can copmlete asynchronously: it's an async version of `HostResolveImportedModule`.
- move the graph traversal logic for the load phase from hosts to ECMA-262: it delegates the actual fetching to hosts with `HostLoadImportedModule`, which is called recursively for each dependency.

Module Records will expose a third method, other than `.Link()` and `.Evaluate()`: `.LoadRequestedModules()`, which is the one responsible of calling `HostLoadImportedModule`.

<details>
<summary>The host algorithm to load/execute a top-level module (for example, using a <code>&lt;script type="module"></code> tag) will now look like this.</summary>

1. (HOST) Fetch the top-level Module Record
2. (HOST) Call `module.LoadRequestedModules()`: 
   - (ECMA-262) For each dependency, call `HostLoadImportedModule(module, specifier)`:
      - (HOST) Fetch the corresponding Module Record and return it.
3. (HOST) When it finishes, call `module.Link()`
   - ... internal steps in ECMA-262 ...
4. (HOST) Call `module.Evaluate()`
   - ... internal steps in ECMA-262 ...

</details>

<details>
<summary>The algorithm for dynamic import will look like this.</summary>

1. (ECMA-262) Call `HostLoadImportedModule(referrer, specifier)`:
    - (HOST) Fetch the corresponding Module Record and return it.
2. (ECMA-262) Call `module.LoadRequestedModules()`: 
   - (ECMA-262) For each dependency, call `HostLoadImportedModule(module, specifier)`:
      - (HOST) Fetch the corresponding Module Record and return it.
3. (ECMA-262) When it finishes, call `module.Link()`
4. (ECMA-262) Call `module.Evaluate()`

</details>


For the Module Blocks proposal, the algorithm to import a module block will look like the above dynamic import algorithm, except that we skip step 1.

For the Import Reflection proposal, the algorithm to load an uninstantiated JS module will just call `HostLoadImportedModule` without then calling the various Module Record methods.

For the Compartments proposal, we will be able to use the new graph loading algorithm by just replacing calls to `HostLoadImportedModule` with calls to an user-provided function.

---

**Some more precise remarks about these HTML changes**

1. ~~The only observable change (observable from HTML, rather than users) is that when you import the same specifier twice from the same file, only the first one causes a call to `HostLoadImportedModule` (unless the first one fails):~~
   ```js
   await import("./dep.js"); // Calls HostLoadImportedModule
   await import("./dep.js"); // Returns the already loaded module
   ```
   ~~In practice this doesn't change much, because `HostImportModuleDynamically`/`HostResolveImportedModule` were already required to return the same Module Record both times.~~

   ~~We still call the host hook multiple times if the first dynamic import fails, because there was an explicit goal of allowing to retry failed dynamic imports (https://github.com/tc39/ecma262/pull/1645).~~
   **EDIT**: For now I reverted this change, since even if it's not observable by HTML it's observable in other places.
2. The first commit is a small refactor only within HTML (it doesn't affect layering) that makes it easier to move the graph traversal to ECMA-262. I believe that it should not have observable effects (I wrote why in the commit description), but if it does I'll find an alternative solution.
3. I had to add an `hostDefined` parameter to `LoadRequestedModules`, which ECMA-262 passes as-is to `HostLoadImportedModule`, so that HTML can propagate the correct request `destination`. I couldn't just store the destination on the referrer Module Record's `[[HostDefined]]` slot because it should only be propagated through static imports and not dynamic ones.
4. On the ECMA-262 side, this refactor doesn't take into account import assertions, because they have not been merged to the main spec yet. I wrote these HTML changes assuming that `HostLoadImportedModule` will receive a `{ [[Specifier]], [[Assertions]] }` record instead of just the specifier, similarly to how the import assertions proposal updated `HostResolveImportedModule` to receive `{ [[Specifier]], [[Assertions]] }`.

---

**Links**
- ECMA-262 PR: https://github.com/tc39/ecma262/pull/2905
- ECMA-262 preview: https://ci.tc39.es/preview/tc39/ecma262/pull/2905
- Slides presented in the TC39 meeting: https://docs.google.com/presentation/d/1RVUE-MENQT8dj2wxvMLMDxg_VoMOwiwNQQged39QIEU/edit?usp=sharing


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/8253/index.html" title="Last updated on Dec 26, 2022, 9:54 AM UTC (75e7bf0)">/index.html</a>  ( <a href="https://whatpr.org/html/8253/085d4d1...75e7bf0/index.html" title="Last updated on Dec 26, 2022, 9:54 AM UTC (75e7bf0)">diff</a> )
<a href="https://whatpr.org/html/8253/infrastructure.html" title="Last updated on Dec 26, 2022, 9:54 AM UTC (75e7bf0)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/8253/085d4d1...75e7bf0/infrastructure.html" title="Last updated on Dec 26, 2022, 9:54 AM UTC (75e7bf0)">diff</a> )
<a href="https://whatpr.org/html/8253/webappapis.html" title="Last updated on Dec 26, 2022, 9:54 AM UTC (75e7bf0)">/webappapis.html</a>  ( <a href="https://whatpr.org/html/8253/085d4d1...75e7bf0/webappapis.html" title="Last updated on Dec 26, 2022, 9:54 AM UTC (75e7bf0)">diff</a> )

domenic commented:
I am skeptical about this change, as the goal of moving graph traversal out of HTML's control and into ES-262's seems likely to be problematic for future extensions we might want to add on the host side. However, I do not have any concrete examples of problems it will cause in mind. Similarly, I am very skeptical about the three proposals you mention, so am unsure about doing work on this side to support them; but, perhaps that's something we can wait for later to figure out.

Can you address how well your changes here would work with proposals such as #6768 and #7996? Similarly, would they allow #4400 or the path-not-taken for #4419? (The latter are not current problems, but knowing whether they would be possible within this framework would help put my mind at ease about whether we are losing important abilities via this change, or not.)

nicolo-ribaudo commented:
Thank you for taking a look at this!

---

### Failed dynamic import should not always be cached (https://github.com/whatwg/html/issues/6768)
`HostLoadImportedModule` does not cache (and does not require caching) failures: when you import the same `(referrer, specifier)` pair twice, if the first import results in a failure it will call `HostLoadImportedModule` also the second time.

Failure caching is only implemented in HTML, and this PR does not change how we can remove it. We can support retrying by either:
- replace step 6 of [fetch a single module script](https://whatpr.org/html/8253/webappapis.html#fetch-a-single-module-script) with "If moduleMap[(url, moduleType)] exists <ins>and is not null</ins>, asynchronously complete this algorithm with moduleMap[url / moduleType], and return.";
- or replace the end of step 13 of [fetch a single module script](https://whatpr.org/html/8253/webappapis.html#fetch-a-single-module-script) with "then <del>set moduleMap[(url, moduleType)] to null,</del> asynchronously complete this algorithm with null, and return.";

### Editorial cleanups to fetching scripts (https://github.com/whatwg/html/issues/7996) <sub>(I love this, it addresses all the doubts I initially had when reading these algorithms :) )</sub>

> - Everything takes a (probably named) onComplete argument, to replace the current "asynchronous completion" paradigm. Yes, callers will have extra indentation. Oh well.

There are a few less places where we need to change this (with this PR there are 39 occurrences of "asynchronously complete" instead of 54 in webappapis.html), but except for that it remains the same task.

> - Make "perform the fetch" an optional named argument and thread it through explicitly.

Are the "perform the fetch" step passes through to _fetch the descendants of and link_, for example in the last step of [fetch an external module script graph](https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-module-script-tree)?
- **Yes** → we need to add a `performTheFetch` "function" to the options bag passed to `module.LoadImportedModule(options)`, which is passed as-is to all the `HostLoadImportedModule` calls.
- **No** → if it only affects the loading of the top-level module, then this PR doesn't affect it because the top-level loading is still done before calling ECMA-262 algorithms.

> - Use modern Fetch's [callbacks](https://fetch.spec.whatwg.org/#fetch-elsewhere). Notably _processResponseConsumeBody_ is better than how we operate on the "body" concept.

We can/should still do it; it makes it clearer when we invoke `HostLoadImportedModules` synchronously/asynchronously (for example, we could invoke it synchronously when returning a module which is already in the `importMap`).

### Yielding to the event loop (in some way) between module evaluations (https://github.com/whatwg/html/issues/4400)

This refactor doesn't change "how much async" `module.Evaluate()` is, so it doesn't affect that proposal. A proper solution would probably require a new host hook in Evaluate like `HostContinueGraphEvaluation(callback)` which must just call `callback` but it could do it either synchronously (default and current behavior) or asynchronously (after yielding to the event loop).

It is possible to "hack a solution" to that with this refactor (or even with the current spec), by replacing step 11 of [HostLoadImportedModule](https://whatpr.org/html/8253/webappapis.html#hostloadimportedmodule(referrer,-modulerequest,-loadstate,-payload)) with something like this:

11. Otherwise,
      1. let completion be Completion { [[Type]]: normal, [[Value]]: _result_'s record, [[Target]]: empty }.
      2. After this algorithm returns, optionally perform the following steps:
             1. <ins>Let _module_ be _result_'s record.</ins>
             2. <ins>Await _module_.LoadRequestedModules(_loadState_).</ins>
             3. <ins>_module_.Link()</ins>
             4. <ins>_module_.Evaluate()</ins>

to evaluate _module_ even if the graph is still loading, just making sure that all its dependencies have already been loaded (11.ii.2). However, it feels "wrong" to hack the expected semantics like this (even if it's compliant to the hook requirements), and I would strongly suggest proposing a first-class mechanism to do it.

### path-not-taken for Top-level await integration and multiple script graphs (https://github.com/whatwg/html/issues/4419)

This is at a different level from this PR, and it touches different parts of the spec. [Execute the script element](https://html.spec.whatwg.org/#execute-the-script-element) would have to await the promise returned by [run the module script](https://html.spec.whatwg.org/#run-a-module-script) instead of just ignoring it.

---

I also took a look at other issues related to module loading:

- This PR fixes https://github.com/whatwg/html/issues/7472, because the refactored dynamic import algorithm on the ECMA-262 side ensures that evaluation of the imported module is delayed by at least one promise tick.
- This PR fixes https://github.com/whatwg/html/issues/3025, because step 9 of [HostLoadImportedModule](https://whatpr.org/html/8253/webappapis.html#hostloadimportedmodule(referrer,-modulerequest,-loadstate,-payload)) creates an TypeError for the _single_ module that fails loading, rather than for the whole module graph (or do we need to _explicitly_ associate the URL with that TypeError?).
- This PR is unrelated to https://github.com/whatwg/html/issues/6761, https://github.com/whatwg/html/issues/6339, and https://github.com/whatwg/html/issues/4594.
- https://github.com/whatwg/html/issues/3744 made me realize that this PR has a bug, because it always sets _referrer_ as `"client"`. I'll push a commit to fix this regression, and it will also fix https://github.com/whatwg/html/issues/3744 (because `HostLoadImportedModule` will compute the referrer without differentiating between static and dynamic import).

Ultimately one goal of this refactor is to not change host capabilities, and the issues fixed by this PR could also be fixed with the existing layering hooks.

---

I also tried a "minimal" alternative approach. Instead of deleting the "recurse into imported modules" algorithm from HTML, `HostLoadImportedModules(referrer, specifier)` can look like this:
1. Let _baseURL_ be the result of resolving (_referrer_, _specifier_).
1. If the module record corresponding to _baseURL_ has already been loaded, synchronously return it.
1. Otherwise,
    1. Fetch the module record corresponding to _baseURL_, and all its recursive dependencies. Wait until it asynchronously completes.
    2. Return the module record.

This is more similar to the existing two hooks: use cases of `HostResolveImportedModule` are covered by steps 1 and 2, while use cases of `HostImportModuleDynamically` are covered by steps 1 and 3.

That leaves most of the HTML algorithms intact, to better show that this refactor shouldn't change host capabilities. If you are interested, I can complete/rebase it and open a parallel PR.

domenic commented:
Thank you so much for the thorough, above-and-beyond analysis! This addresses my concerns. At this point I'm somewhat eager for the refactor to land, and no longer think we need to wait for a TC39 proposal that needs it.

> Are the "perform the fetch" step passes through to fetch the descendants of and link, for example in the last step of fetch an external module script graph?

Yes, that is the intent. (With the "is top-level" flag being used to allow branching behavior as necessary.)

I guess we can make this work by using the [[HostDefined]] field of the ModuleLoadState Record?

Speaking of that, I noticed that it says loadState is undefined for dynamic import(). Is that a problem? Is it something we could change in the future, if e.g. we wanted to change the fetch destination for dynamic import(), or change the "perform the fetch" steps?

> It is possible to "hack a solution" to that with this refactor (or even with the current spec), by replacing step 11 of HostLoadImportedModule with something like this:

Well, with the current spec, we could just choose not to call the top-level Evaluate(). Instead we would Evaluate() portions of the module graph, as the host determines would be best. This seems possible because the host has the graph-traversal power, so it can delay the scheduling of Evaluate() on various subgraphs as necessary.

It sounds like maybe we have another mechanism for exerting that control, by basically inserting host-defined top-level awaits? Maybe that's enough; back when we first raised #4400, top-level await wasn't integrated, so that option wasn't available.

> This PR fixes Behavior of cyclical dynamic import() is unclear #7472, because the refactored dynamic import algorithm on the ECMA-262 side ensures that evaluation of the imported module is delayed by at least one promise tick.
>
> This PR fixes Track the source of fetching errors in module graphs #3025, because step 9 of HostLoadImportedModule creates an TypeError for the single module that fails loading, rather than for the whole module graph (or do we need to explicitly associate the URL with that TypeError?).

This is great news!

> it will also fix import() should use referring script's base URL as the referrer #3744 (because HostLoadImportedModule will compute the referrer without differentiating between static and dynamic import).

Fixing this would be a normative change, and thus require web platform tests, and (depending on what the tests show) possibly implementer agreement. So it might be best to hold off on that for now?

> I also tried a "minimal" alternative approach. Instead of deleting the "recurse into imported modules" algorithm from HTML, HostLoadImportedModules(referrer, specifier) can look like this:

Oh, this is very helpful. No need to write up a full PR with this, but knowing that we have that option in our back pocket if necessary is really helpful.

nicolo-ribaudo commented:
> Speaking of that, I noticed that it says loadState is undefined for dynamic import(). Is that a problem? Is it something we could change in the future, if e.g. we wanted to change the fetch destination for dynamic import(), or change the "perform the fetch" steps?

We need that LoadStateRecord.[[HostDefined]] record to differentiate between "the current graph loading process" (static imports) and "future graph loading processes" (dynamic imports). Each loading process has comes from a separate LoadRequestedModules call, and thus can have its own loading-process-specific data.

If we wanted to pass data from the referrer module to the imported module, regardless of when its imported, then we can store it on the module script struct, which HostLoadImportedModule can access using referrer.[[HostDefined]] (this was my original attempt, until I realized that the destination for dynamic import is always `"script"` and thus I introduced LoadStateRecord.[[HostDefined]]).

> Fixing this would be a normative change, and thus require web platform tests, and (depending on what the tests show) possibly implementer agreement. So it might be best to hold off on that for now?

Ok, I'll keep it for a future PR. It's fun that with this refactor is easier to fix it than to not, and I'll have to explicitly unset the referrer for dynamic imports :)

nicolo-ribaudo commented:
@domenic I have rebased this on top of the asynchronous completions removal, fixed the referrer handling to restore the old behavior and properly fixed the "perform the fetch" steps propagation.

I plan to continue working on #7996 (which obviously has conflicts with this PR), but I would still love a review of this one in parallel.

annevk commented on source:
```diff
@@ -2891,6 +2893,7 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
      <li>The <dfn data-x-href="https://tc39.es/ecma262/#sec-detacharraybuffer">DetachArrayBuffer</dfn> abstract operation</li>
      <li>The <dfn data-x-href="https://tc39.es/ecma262/#sec-enumerableownproperties">EnumerableOwnProperties</dfn> abstract operation</li>
      <li>The <dfn data-x-href="https://tc39.es/ecma262/#sec-finishdynamicimport">FinishDynamicImport</dfn> abstract operation</li>
+     <li>The <dfn data-x-href="https://ci.tc39.es/preview/tc39/ecma262/pull/2905/#sec-FinishLoadImportedModule">FinishLoadImportedModule</dfn> abstract operation</li>
```
Is the plan to land this with these kind of links? Or will we wait for the JS PR to land and then adjust this?

nicolo-ribaudo replied:
We can land the JS PR first and then adjust the links here, but the JS PR is blocked on this one being reviewed.

annevk replied:
This might slightly clash with @domenic's #8075 btw. In particular the _referrer_-related aspects look like they conflict.

nicolo-ribaudo replied:
Just at a diff level, right? Because to merge both PRs we can do
1. If _referrer_ is a Script Record or a Module Record, let _referencing script_ be _referrer_.[[HostDefined]]
2. Else, let _referencing script_ be null.

domenic replied:
I'll mark "do not merge yet" blocking on landing the JS PR first. I didn't realize this needed review; will do that now.

annevk commented on source:
```diff
@@ -93284,18 +93233,11 @@ document.querySelector("button").addEventListener("click", bound);
-       descendants of and link</span> <var>result</var> given <var>settings object</var>,
-       <var>destination</var>, <var>visited set</var>, and with an empty algorithm.</p></li>
-      </ol>
+      <p>If <var>result</var> is not null, optionally p <span data-x="fetch the descendants of and
```
optionally p?

annevk commented on source:
```diff
@@ -93284,18 +93233,11 @@ document.querySelector("button").addEventListener("click", bound);
+      <var>settings object</var>, <var>destination</var>, and an empty algorithm.</p>
 
-      <p class="note">Generally, performing these steps will be beneficial for performance, as it
+      <p class="note">Generally, performing this steps will be beneficial for performance, as it
```
"these" was correct.

nicolo-ribaudo replied:
"this step", since there is just one now (there were two before this PR)

annevk commented on source:
```diff
@@ -93514,208 +93428,80 @@ document.querySelector("button").addEventListener("click", bound);
 
-   <li><p>Let <var>options</var> be the <span>descendant script fetch options</span> for <var>module
-   script</var>'s <span data-x="concept-script-script-fetch-options">fetch options</span>.</p></li>
+   <li><p>Let <var>state</var> be Record { [[ParseError]]: null, [[Destination]]:
```
Nit: xref Record as we do elsewhere.

annevk reviewed (commented):
Minor drive-by nits, deferring to @domenic for overall review.

domenic commented on source:
```diff
@@ -2901,10 +2904,9 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+     <li>The <dfn data-x="js-HostLoadImportedModule" data-x-href="https://ci.tc39.es/preview/tc39/ecma262/pull/2905/#sec-HostLoadImportedModule">HostLoadImportedModule</dfn> abstract operation</li>
      <li>The <dfn data-x="js-HostMakeJobCallback" data-x-href="https://tc39.es/ecma262/#sec-hostmakejobcallback">HostMakeJobCallback</dfn> abstract operation</li>
      <li>The <dfn data-x="js-HostPromiseRejectionTracker" data-x-href="https://tc39.es/ecma262/#sec-host-promise-rejection-tracker">HostPromiseRejectionTracker</dfn> abstract operation</li>
-     <li>The <dfn data-x="js-HostResolveImportedModule" data-x-href="https://tc39.es/proposal-import-assertions/#sec-hostresolveimportedmodule">HostResolveImportedModule</dfn> abstract operation</li>
```
Hmm, will proposal-import-assertions be updated at the same time? We had readers confused recently in https://github.com/whatwg/html/issues/8262 by the HTML spec not referring to a version of HostResolveImportedModule that worked well with import assertions; I hope we can avoid regressing on that.

domenic commented on source:
```diff
@@ -92925,10 +92927,9 @@ document.querySelector("button").addEventListener("click", bound);
-  statements encountered throughout the graph.</p>
+  graph</span>, or <span data-x="fetch a module worker script graph">fetching a module worker script
+  graph</span>, but not for the fetches resulting from <code data-x="">import</code> statements
+  encountered throughout the graph or from <code data-x="">import()</code> expressions.</p>
```
I guess this change is fine, since the two consumers of "is top-level" (modulepreload and service workers) don't care about whether dynamic import() triggers that code path or not?

nicolo-ribaudo replied:
Exactly, that flag only has effect if:
- the destination is `"worker"` or `"serviceworker"`, but for dynamic imports it's always `"script"`
- if there are custom perform the fetch steps, but they are never present for dynamic imports

domenic commented on source:
```diff
@@ -93380,7 +93319,7 @@ document.querySelector("button").addEventListener("click", bound);
   other:</p>
 
-  <svg id="module-script-fetching-diagram" viewBox="0 0 1131 256" style="width: 100%;" role="img" aria-label="Fetch an external module script, fetch an import() module script graph, fetch a modulepreload module script graph, fetch an inline module script graph, and fetch a module worker script graph all call fetch the descendants of and link a module script. It, in turn, calls fetch the descendan ...
+  <svg id="module-script-fetching-diagram" viewBox="0 0 1131 256" style="width: 100%;" role="img" aria-label="Fetch an external module script, fetch a modulepreload module script graph, fetch an inline module script graph, and fetch a module worker script graph all call fetch the descendants of and link a module script.">
```
The coordinates of the others boxes in the diagram need updating to not have a blank box in the middle, and whitespace at the bottom :). Hopefully the math is relatively easy to figure out...

domenic commented on source:
```diff
@@ -93514,208 +93428,80 @@ document.querySelector("button").addEventListener("click", bound);
-    object</span>, <var>visited set</var>, <var>module script</var>'s <span
-    data-x="concept-script-base-url">base URL</span>, and the following steps given
-    <var>result</var>:</p>
+    <p><span>Upon rejection</span> of <var>loading promise</var>, run the
```
Usually we put "upon fulfillment" steps before rejection ones.

domenic commented on source:
```diff
@@ -93514,208 +93428,80 @@ document.querySelector("button").addEventListener("click", bound);
+      <p class="note">This step will recursively call <span data-x="js-Link">Link</span> on all of
+      the module's unlinked dependencies.</p>
+
+      <p>If this throws an exception, set <var>module script</var>'s <span
```
```suggestion
      <p>If this throws an exception, catch it, and set <var>module script</var>'s <span
```

domenic commented on source:
```diff
@@ -93514,208 +93428,80 @@ document.querySelector("button").addEventListener("click", bound);
-    <p class="note">The fetches performed by the <span>internal module script graph fetching
-    procedure</span> are performed in parallel to each other.</p>
+     <!-- loading error -->
+     <li><p>Otherwise, run <var>onComplete</var> give null.</p></li>
```
```suggestion
     <li><p>Otherwise, run <var>onComplete</var> given null.</p></li>
```

domenic commented on source:
```diff
@@ -93514,208 +93428,80 @@ document.querySelector("button").addEventListener("click", bound);
-
-    <p class="note">The fetches performed by the <span>internal module script graph fetching
-    procedure</span> are performed in parallel to each other.</p>
+     <!-- loading error -->
```
Probably this should be an (expanded) `<p class="note">` so that readers of the rendered spec know what's going on, not just readers of the source file.

domenic commented on source:
```diff
@@ -95247,33 +95007,38 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
 
-  <h6><dfn>HostImportModuleDynamically</dfn>(<var>referencingScriptOrModule</var>,
-  <var>moduleRequest</var>, <var>promiseCapability</var>)</h6>
+  <h6><dfn>HostLoadImportedModule</dfn>(<var>referrer</var>, <var>moduleRequest</var>,
```
```suggestion
  <h6 id="the-hostloadimportedmodule-implementation"><dfn>HostLoadImportedModule</dfn>(<var>referrer</var>, <var>moduleRequest</var>,
```

domenic commented on source:
```diff
@@ -95247,33 +95007,38 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
+
+    <p class="note">
+     <var>loadState</var> is undefined when the current fetching process has been initiated by a
+     dynamic <code data-x="">import()</code> call.
```
```suggestion
     dynamic <code>import()</code> call.
```

domenic commented on source:
```diff
@@ -95247,33 +95007,38 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
+    object</span> implements <code>WorkletGlobalScope</code> or <code>ServiceWorkerGlobalScope</code>
+    and <var>loadState</var> is undefined, then:</p>
+
+    <p class="note">
```
Nit: no newlines inside `<p>`s.

domenic commented on source:
```diff
@@ -95305,104 +95073,76 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
      <li><p>Assert: Neither <var>base URL</var> nor <var>fetch options</var> is null, as
```
```suggestion
     <li><p>Assert: neither <var>base URL</var> nor <var>fetch options</var> is null, as
```

domenic commented on source:
```diff
@@ -95305,104 +95073,76 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
 
-   <li><p>Let <var>moduleMap</var> be <var>settings object</var>'s <span
-   data-x="concept-settings-object-module-map">module map</span>.</p></li>
+   <li><p>Let <var>destination</var> be "script".</p></li>
```
```suggestion
   <li><p>Let <var>destination</var> be "<code data-x="">script</code>".</p></li>
```

domenic commented on source:
```diff
@@ -95305,104 +95073,76 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
-   <li><p>Let <var>url</var> be the result of <span data-x="resolve a module specifier">resolving a
-   module specifier</span> given <var>base URL</var> and
-   <var>moduleRequest</var>.[[Specifier]].</p></li>
+   <li><p>If <var>loadState</var> is not undefined, set <var>destination</var> to
```
```suggestion
   <li><p>If <var>loadState</var> is not undefined, then set <var>destination</var> to
```

domenic commented on source:
```diff
@@ -95305,104 +95073,76 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
+    <p><span>Fetch a single imported module script</span> given <var>settings object</var>,
+    <var>base URL</var>, <var>destination</var>, <var>fetch options</var>, <var>fetch
+    referrer</var>, <var>moduleRequest</var>, and with the following steps given <var>module
+    script</var>:</p>
```
Let's use _moduleScript_ here

domenic commented on source:
```diff
@@ -95305,104 +95073,76 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
-   <li><p>Assert: <var>resolved module script</var> is a <span>module script</span> (i.e., is not
-   null or "<code data-x="">fetching</code>").</p></li>
+      <ol>
+       <li><p>Let <var>parse error</var> be <var>module script</var>'s <span
```
Let's use _parseError_ here.

domenic commented on source:
```diff
@@ -95305,104 +95073,76 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
+     <var>moduleRequest</var>, <var>payload</var>, <var>completion</var>).</p></li>
+    </ol>
+
+    <p>If <var>loadState</var> is not undefined and <var>loadState</var>.[[PerformTheFetch]] is not
```
As noted in #8328, let's try to avoid modifying the call site from so far away.

domenic commented on source:
```diff
@@ -2891,6 +2893,7 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
      <li>The <dfn data-x-href="https://tc39.es/ecma262/#sec-detacharraybuffer">DetachArrayBuffer</dfn> abstract operation</li>
      <li>The <dfn data-x-href="https://tc39.es/ecma262/#sec-enumerableownproperties">EnumerableOwnProperties</dfn> abstract operation</li>
      <li>The <dfn data-x-href="https://tc39.es/ecma262/#sec-finishdynamicimport">FinishDynamicImport</dfn> abstract operation</li>
+     <li>The <dfn data-x-href="https://ci.tc39.es/preview/tc39/ecma262/pull/2905/#sec-FinishLoadImportedModule">FinishLoadImportedModule</dfn> abstract operation</li>
```
Maybe consider a more grammatical name, like "FinishLoadingImportedModule".

domenic reviewed (commented):
Generally looks good. I think we have some editorial work to do, which we'll probably make harder for ourselves due to the other concurrent module-related PRs. But I have no general objections.

nicolo-ribaudo commented on source:
```diff
@@ -95302,107 +95078,79 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
+    <var>base URL</var>, <var>destination</var>, <var>fetch options</var>, <var>fetch
+    referrer</var>, <var>moduleRequest</var>, and <var>onSingleFetchComplete</var> as defined below.
+    If <var>loadState</var> is not undefined and <var>loadState</var>.[[PerformTheFetch]] is not
+    null, pass <var>loadState</var>.[[PerformTheFetch]] along as well.</p>
```
I'm assuming that https://github.com/whatwg/html/pull/8328 lands first.

nicolo-ribaudo commented:
Thanks for the review! I plan to prepare a PR in the import assertions proposal, but I don't know how long it will take to get merged. I added a `<p class="note">` to the `HostLoadImportedModule` implementation explaining how it expects the import assertions proposal to update the hook's signature, to avoid confusion about mismatched types.

nicolo-ribaudo commented:
I rebased this on top of the import maps changes, hopefully I didn't miss anything.

While doing so, I found a bug in the workaround to preserve https://github.com/whatwg/html/issues/3744: instead of checking if this import was _directly_ caused by `import()`, I'm checking if it's caused by `import()` even indirectly (for example, by a static import in a dynamically imported file). I'll push a fix soon.

domenic commented on source:
```diff
@@ -93439,12 +93385,9 @@ document.querySelector("button").addEventListener("click", bound);
-   destination "<code data-x="">script</code>", <var>visited set</var>, and
-   <var>onComplete</var>.</p></li>
+   <li><p><span data-x="fetch the descendants of and link a module script">Fetch the descendants of
+   and link</span> <var>script</var>, given <var>settings object</var>, the destination "<code
```
```suggestion
   and link</span> <var>script</var>, given <var>settings object</var>, "<code
```

while we're in the area

domenic commented on source:
```diff
@@ -93502,7 +93445,7 @@ document.querySelector("button").addEventListener("click", bound);
   other:</p>
 
-  <svg id="module-script-fetching-diagram" viewBox="0 0 1131 256" style="width: 100%;" role="img" aria-label="Fetch an external module script, fetch an import() module script graph, fetch a modulepreload module script graph, fetch an inline module script graph, and fetch a module worker script graph all call fetch the descendants of and link a module script. It, in turn, calls fetch the descendan ...
+  <svg id="module-script-fetching-diagram" viewBox="0 0 941 256" style="width: 80%; max-width: 1024px" role="img" aria-label="Fetch an external module script, fetch a modulepreload module script graph, fetch an inline module script graph, and fetch a module worker script graph all call fetch the descendants of and link a module script.">
```
Why change the `style=""`?

domenic replied:
Diagram still has whitespace at the bottom which show be removed, so that we have roughly the same margin around everything.

domenic commented on source:
```diff
@@ -96160,107 +95932,108 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
-    <var>referencing script</var>, <var>settings object</var>, <var>fetch options</var>, and with
-    the following steps given <var>result</var>:</p>
+     <li>
+      <p>If none of the following conditions is true</p>
```
Let's do this typgraphy/wording:

```
If neither of the following conditions are true:

- A; or
- B,

then ...
```

domenic commented on source:
```diff
@@ -96160,107 +95932,108 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
-   <li><p>Let <var>resolved module script</var> be <var>moduleMap</var>[(<var>url</var>,
-   <var>moduleType</var>)]. (This entry must <span data-x="map exists">exist</span> for us to have
-   gotten to this point.)</p></li>
+       <li><p>Let <var>completion</var> be the <span>Completion Record</span> { [[Type]]: throw,
```
Here and elsewhere, it looks like the latest 262 conventions are to omit the "the". (I filed https://github.com/whatwg/html/issues/8364 about existing usage.)

domenic commented on source:
```diff
@@ -96160,107 +95932,108 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
+    <ol>
+     <li><p>If <var>moduleScript</var> is null, then let <var>completion</var> be the
+     <span>Completion Record</span> { [[Type]]: throw, [[Value]]: a new <code>TypeError</code>,
+     [[Target]]: empty }.</p></li>
```
We try to treat "let" as block-scoped, so it's not great to declare them inside ifs (even single-line ifs). So maybe make substep 1 establish the default (currently set on step 3), and then the other substeps can be changed from "let" to "set".

nicolo-ribaudo replied:
I ended up initializing it as `null` and keep the Completion Record creations where they are, it felt strange to initialize it to a record representing success and then later be like "oh actually, it was an error".

domenic reviewed (approved):
LGTM with nits. I would prefer if import assertions were merged into the JS spec as close as possible to your refactoring, to avoid the temporary confusion. But this is reasonable for now.

nicolo-ribaudo commented:
The ECMA-262 side of this refactor has been merged, I'll rebase this PR now!

nicolo-ribaudo commented:
Note: https://github.com/tc39/proposal-import-assertions/pull/124 has not been merged yet, but the HTML implementation of `HostLoadImportedModule` has a note explaining how it expects the import assertions proposal to update its signature. I will open a PR to remove that note once the import assertions PR is merged.
